### PR TITLE
[Bug] Add explicit format preserver

### LINKED ISSUE #2418: [Bug] Conversion from toml to ndjson breaks mail action message for detection rules
**Describe the bug**
When converting a mail connection action with newlines from ndjson to toml, rule.TOMLRule.save_toml() removes newlines. This causes the conversion back to ndjson to fail (output contains wrong formatting or conversion fails due to toml parse errors). 

Mustache supports the use of backslashes and double-spaces as newline characters. Headers (## my header) get automatic newlines.  The returned TOMLRule object contains the original correct formatting, so it seems to be the writing through .save_toml() that causes issues. More specifically, I believe everything goes well in .save_toml() until the call to rule_formatter.toml_write() which removes the newline characters and double spaces. 

**To Reproduce**
1. Export a detection rule with a mail connector action with Moustache newlines (either double space + enter or backslash + enter) or headers (### My header + newline). Characters are preserved in the ndjson. 
2. Convert to TOML using create-rule() with the exported ndjson path as input and required_only set to False.
3. The formatting of message field in the toml output is destroyed or conversion fails.

**Expected behavior**
When converting from ndjson to TOML, formatting of the message field should be maintained.

### PR DESCRIPTION
**Pull request by Mikaayenson**
## Issues

Resolves #2418

## Summary

- Adds a new method `preserve_formatting_for_fields` that can traverse nested dictionaries and mark as a `NonformattedField` so that newlines and whitespaces are preserved. The method is meant to be abstracted enough to support preserving any field by passing in as a dotted notation.
- Preserves the `action.params.message` field.

## Testing
- Run the importer with the sample file: `python -m detection_rules import-rules /Users/stryker/Downloads/rules_export_mail.ndjson` [rules_export_mail.ndjson.txt](https://github.com/elastic/detection-rules/files/14839897/rules_export_mail.ndjson.txt). The `message` field should preserve newlines in the action.

### CODE REVIEW

File: detection_rules/rule_formatter.py [hunk 1]
```diff
@@ -74,6 +74,31 @@ class NonformattedField(str):
...
+
+    def apply_preservation(target: OrderedDict, keys: list):
+        """Apply NonformattedField preservation based on keys path."""
+        for key in keys[:-1]:
+            # Iterate to the  key, diving into nested dictionaries
+            if key in target and isinstance(target[key], dict):
+                target = target[key]
+            else:
+                # Cannot preserve formatting for missing or non-dict intermediate
+                return
```
Comment (eric-forte-elastic) [inline detection_rules/rule_formatter.py:88]: Should we warn here that the preservation failed?

File: detection_rules/rule_formatter.py [hunk 2]
```diff
@@ -74,6 +74,31 @@ class NonformattedField(str):
...
 
 
+def preserve_formatting_for_fields(data: OrderedDict, fields_to_preserve: list) -> OrderedDict:
+    """Preserve formatting for specified nested fields in an action."""
+
+    def apply_preservation(target: OrderedDict, keys: list):
+        """Apply NonformattedField preservation based on keys path."""
+        for key in keys[:-1]:
+            # Iterate to the  key, diving into nested dictionaries
+            if key in target and isinstance(target[key], dict):
```
Comment (eric-forte-elastic) [inline detection_rules/rule_formatter.py:84]:
```suggestion
            # Iterate to the key, diving into nested dictionaries
            if key in target and isinstance(target[key], dict):
```

File: detection_rules/rule_formatter.py [hunk 3]
```diff
@@ -74,6 +74,31 @@ class NonformattedField(str):
     """Non-formatting class."""
 
 
+def preserve_formatting_for_fields(data: OrderedDict, fields_to_preserve: list) -> OrderedDict:
+    """Preserve formatting for specified nested fields in an action."""
+
+    def apply_preservation(target: OrderedDict, keys: list):
```
Comment (eric-forte-elastic) [inline detection_rules/rule_formatter.py:80]:
```suggestion
    def apply_preservation(target: OrderedDict, keys: list) -> None:
```

File: detection_rules/rule_formatter.py (hunk 1 above)
Comment (Mikaayenson) [reply detection_rules/rule_formatter.py:88]: This should never happen TBH. It's not a dynamic field.

**Review by eric-forte-elastic — APPROVED** [review]
:green_circle: Testing Manual review, looks good to me! :+1:

<details><summary>Details</summary>
<p>

```shell
detection-rules on  2418-bug-conversion-from-toml-to-ndjson-breaks-mail-action-message-for-detection-rules [$!?] is  v0.1.0 via  v3.12.2 (detection-rules-build) on  eric.forte 
❯ python -m detection_rules import-rules ~/Downloads/rules_export_mail.ndjson

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

[+] Building rule for /home/forteea1/Code/clean_mains/detection-rules/rules/test_mail_rule.toml
alert_suppression: 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): 
author (required)  (multi, comma separated): Elastic
building_block_type: 
data_view_id: 
enabled: 
event_category_override: 
exceptions_list (multi, comma separated): 
false_positives (multi, comma separated): 
filters (multi, comma separated): 
investigation_fields: 
license: 
note: 
references (multi, comma separated): 
related_integrations (multi, comma separated): 
required_fields (multi, comma separated): 
risk_score_mapping (multi, comma separated): 
rule_name_override: 
setup: 
severity_mapping (multi, comma separated): 
tags (multi, comma separated): 
add mitre tactic? [y/N]: 
throttle: 
tiebreaker_field: 
timeline_id: 
timeline_title: 
timestamp_field: 
timestamp_override: 

```

image


</p>
</details>

**Comment by shashank-elastic** [comment]
🟢 Testing Looks Good

<details>

```
❯ python -m detection_rules import-rules ~/Downloads/rules_export_mail.ndjson 

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

[+] Building rule for /Users/shashankks/elastic_workspace/detection-rules/rules/test_mail_rule.toml
alert_suppression: 
author (required)  (multi, comma separated): Elastic
building_block_type: 
data_view_id: 
enabled: 
event_category_override: 
exceptions_list (multi, comma separated): 
false_positives (multi, comma separated): 
filters (multi, comma separated): 
investigation_fields: 
license: 
note: 
references (multi, comma separated): 
related_integrations (multi, comma separated): 
required_fields (multi, comma separated): 
risk_score_mapping (multi, comma separated): 
rule_name_override: 
setup: 
severity_mapping (multi, comma separated): 
tags (multi, comma separated): 
add mitre tactic? [y/N]: 
throttle: 
tiebreaker_field: 
timeline_id: 
timeline_title: 
timestamp_field: 
timestamp_override: 
```
Rule File created 
image

</details>

**Review by shashank-elastic — APPROVED** [review]

File: detection_rules/rule_formatter.py [hunk 4]
```diff
@@ -74,6 +74,31 @@ class NonformattedField(str):
...
 
 
+def preserve_formatting_for_fields(data: OrderedDict, fields_to_preserve: list) -> OrderedDict:
+    """Preserve formatting for specified nested fields in an action."""
+
+    def apply_preservation(target: OrderedDict, keys: list) -> None:
+        """Apply NonformattedField preservation based on keys path."""
+        for key in keys[:-1]:
+            # Iterate to the key, diving into nested dictionaries
+            if key in target and isinstance(target[key], dict):
```
Comment (brokensound77) [inline detection_rules/rule_formatter.py:84]: I don't think we have any list of objects that fall into this category, so should be good for this patch (and hopefully a refactor or replacement of this process soon)

**Review by brokensound77 — APPROVED** [review]
Nice job finding a viable patch for this.

Don't feel great about having to hard code exceptions like this, but it should be fine for a hopefully not too distant refactor or replacement.

LGTM 🚀

**Pull request merged**